### ISSUE TITLE
[Proposal] Stand-alone Where Clauses

### ISSUE THREAD

**Issue by timhuff** [knex/knex#1893]
I wanted to bounce a feature idea off the community here.

I think it'd be nice to have standalone where clauses that you could pass around. My proposal for the API would be:
```javascript
var whereClause = knex.where({state: 'CA'});
knex.select().from('users').where(whereClause);
```

The purpose here would be the ability to pass where clauses into functions. So you could make a function that looks like this:
```javascript
var getUsers = function(whereClause){
    return knex.select().from('users').where(whereClause);
};
```
and then call it with:
```javascript
getUsers(knex.where({state:'CA'}));
```

(Obviously, this would be a useless abstraction - just trying to keep the example simple)

I, personally, have a use case where this functionality would be useful. In the example above, the solution might be to just `getUsers().where({state:'CA'})` but for my use case, the where clause is actually applied to a subquery.

I currently solve the problem by defining the function like so:
```javascript
var getUsers = function(userSelector){
    return knex.select().from('users').whereIn('username', userSelector);
};
```
and then executing it like so:
```javascript
getUsers(knex.select('username').from('users').where({state: 'CA'})
```

(I realize that this solution has potential drawbacks but, for my use case, they're not significant)

**Comment by timhuff** [comment]
I've written a piece of CoffeeScript that enables this functionality. After looking at the code base, it appears that implementation would be really straight-forward.

```coffeescript
	knex.queryBuilder().constructor.prototype.appendWhere = (clause)->
		Builder = knex.queryBuilder().constructor
		if !(clause instanceof Builder)
			throw new Error 'whereClase() expects a query builder.'
		if this._single.table
			throw new Error 'You passed a query builder into whereClause() that has a table specified for it'
		for statement in clause._statements
			if statement.grouping != 'where'
				throw new Error 'You passed a query builder into whereClause() that had a non-where clause on it.'
			this._statements.push statement
		return this

```

**Comment by elhigu** [comment]
@timhuff I didn't knew that `.where()` didn't accept query builder as parameter... Indeed it sounds reasonable and intuitive, but it also is a bit strange since now we are passing "partial" subquery to `.where()` instead of complete subquery how it is usually used.

**Comment by timhuff** [comment]
I'm not sure what you mean by "instead of complete subquery how it is usually used". Are you referring to `.whereIn()`?

**Comment by timhuff** [comment]
@elhigu So I hate to say it because I spent a decent amount of time on that pull request but I just realized that this might be unnecessary. I mean, it's decent syntactic sugar but, ultimately, the same effect can be achieved like so:
```javascript
var whereClause = function(){this.where({state: 'CA'})};
knex.select().from('users').where(whereClause);
```
Which is a little verbose, but I write in coffeescript so it'd just be
```coffeescript
whereClause = ->@where state: 'CA'
knex.select().from('users').where(whereClause)
```

**Comment by timhuff** [comment]
I guess the drawback of the above method would be that it's not very extensible. You can't pass around a `whereClause` function while tagging on `.where` chains.

**Comment by elhigu** [comment]
@timhuff 

> I'm not sure what you mean by "instead of complete subquery how it is usually used". Are you referring to .whereIn()?

`whereIn`is one of the places and there are many more places where you can pass subquery... for example:

```js
.where('id', subQuery) 
.raw('hey lets put subquery here ?', [subQuery])
.select([subQuery, 'columnName'])
.update({ column: subQuery })
```

About passing function to the `.where` that was my first thought when I read this issue that you could use them, but there are some differences and indeed they are syntactically that nice.

If you use functions combining where clauses every part will always force parenthesis around them. 

So functions are not really merging where clauses but makes kind of graph that is resolved later on when query builder is compiled... quite subtle difference though...

**Comment by timhuff** [comment]
I agree. It's a subtle difference but I do think the pull request remains a worthy improvement to the code base. As we've demonstrated, it's not new functionality but it makes the API cleaner, IMO. I think the difference is most significant when building up where clauses. Here's the before/after of that use-case:

Before:
```javascript
var scopeToActiveUsers = function(whereClause){
    return function(){ this.where({active: true}).where(whereClause)};
};
var whereClause = function(){ this.where({state: 'CA'})};
var scopedWhereClause = scopeToActiveUsers(whereClause);
knex.select().from('users').where(scopedWhereClause);
```

After:
```javascript
var scopeToActiveUsers = function(whereClause){
    return whereClause.where({active: true});
};
var whereClause = knex.where({state: 'CA'});
var scopedWhereClause = scopeToActiveUsers(whereClause);
knex.select().from('users').where(whereClause);
```

For my use case, this difference is exacerbated by conditional where clauses, where I want to scope a query only under certain conditions.

**Comment by mercmobily** [comment]
Thanks for proposing the ability to pass query builders to `.where()` for composing WHERE clauses.

I don't think we should not add this feature

Your use case is valid - composing and reusing WHERE clauses would be cleaner with this syntax:

```javascript
var whereClause = knex.where({state: 'CA'});
knex.select().from('users').where(whereClause);
```

Why am I against it?


1. **Existing solution works well** - The function approach already provides this capability:
```javascript
const whereClause = function() { this.where({state: 'CA'}) };
knex.select().from('users').where(whereClause);
```

2. **API confusion** - As @elhigu noted, everywhere else in Knex where you pass a builder (whereIn, select, update, raw), it's a complete subquery, not a partial clause. This would be inconsistent. This is my main gripe with this request.

3. **Implementation complexity** - As @wubzz pointed out, we'd need to:
   - Prevent non-WHERE builders from being passed
   - Handle insert/update/delete builders that would break
   - Extract and wrap the WHERE clause correctly
   - Deal with numerous edge cases

4. **Limited benefit** - The syntactic improvement is minor compared to the complexity and potential for misuse

5. **Parentheses behaviour differs** - Functions create proper parenthetical grouping which is often desirable for WHERE clause composition. Direct builder merging would lose this.

**The function approach is actually better**

The existing function approach:
- Provides proper parenthetical grouping
- Is explicit about what's happening
- Works today without changes
- Can't be misused with wrong builder types
- Is consistent with Knex's existing patterns

**Your specific use case**

For conditional WHERE clause building, the current approach works well:
```javascript
let query = knex.select().from('users');
if (condition1) query = query.where({state: 'CA'});
if (condition2) query = query.where({active: true});
```

**Conclusion**

While the syntax would be slightly cleaner, the implementation complexity, API inconsistency, and potential for misuse outweigh the minor benefits. The existing function approach is the right solution.

**Related issues**

- #2526 - Similar request for WHERE clause extraction
- #2673 - Related PR for clause generation
- General query composition patterns in Knex
